src/main/java/io/github/thunderz99/cosmos/CosmosDatabase.java

```diff
@@ -89,6 +98,294 @@ public CosmosDocument create(String coll, Object data, String partition) throws
+    }
+
+    /**
+     * Upsert batch documents in a single transaction.
+     * Note: the maximum number of operations is 100.
+     *
+     * @param coll      collection name
+     * @param data      data object
+     * @param partition partition name
+     * @return CosmosDocument instances
+     * @throws Exception CosmosException
+     */
+    public List<CosmosDocument> batchUpsert(String coll, List<?> data, String partition) throws Exception {
+        doCheckBeforeBatch(coll, data, partition);
+
+        var partitionKey = new com.azure.cosmos.models.PartitionKey(partition);
+        var container = this.clientV4.getDatabase(db).getContainer(coll);
+        CosmosBatch batch = CosmosBatch.createCosmosBatch(partitionKey);
+        data.forEach(it -> {
+            var map = JsonUtil.toMap(it);
+            map.put(Cosmos.getDefaultPartitionKey(), partition);
+            batch.upsertItemOperation(map);
+        });
+
+        return doBatchWithRetry(container, batch);
+    }
+
+    /**
+     * Delete batch documents in a single transaction.
+     * Note: the maximum number of operations is 100.
+     *
+     * @param coll      collection name
+     * @param data      data object
+     * @param partition partition name
+     * @return CosmosDocument instances (only id)
+     * @throws Exception CosmosException
+     */
+    public List<CosmosDocument> batchDelete(String coll, List<?> data, String partition) throws Exception {
+        doCheckBeforeBatch(coll, data, partition);
+
+        var partitionKey = new com.azure.cosmos.models.PartitionKey(partition);
+        var container = this.clientV4.getDatabase(db).getContainer(coll);
+        CosmosBatch batch = CosmosBatch.createCosmosBatch(partitionKey);
+
+        var ids = new ArrayList<String>();
+        data.stream().map(CosmosDatabase::getId).filter(ObjectUtils::isNotEmpty).forEach(it -> {
+            ids.add(it);
+            batch.deleteItemOperation(it);
+        });
+
+        doBatchWithRetry(container, batch);
+
+        return ids.stream().map(it -> {
+            var doc = new Document(JsonUtil.toJson(Map.of("id", it)));
+            return new CosmosDocument(doc.toObject(JSONObject.class));
+        }).collect(Collectors.toList());
+    }
+
+    private static String getId(Object object) {
+        String id;
+        if (object instanceof String) {
+            id = (String) object;
+        } else {
+            var map = JsonUtil.toMap(object);
+            id = map.get("id").toString();
+        }
+        return id;
+    }
+
+    private static void doCheckBeforeBatch(String coll, List<?> data, String partition) {
+        Checker.checkNotBlank(coll, "coll");
+        Checker.checkNotBlank(partition, "partition");
+        Checker.checkNotEmpty(data, "create data " + coll + " " + partition);
+
+        checkBatchMaxOperations(data);
+        checkValidId(data);
+    }
+
+    private static void doCheckBeforeBulk(String coll, List<?> data, String partition) {
```

Comment:
Please change private to package and add unit test for this method (include boundary checks).